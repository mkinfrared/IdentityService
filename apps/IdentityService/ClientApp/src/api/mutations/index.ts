export * from "./loginMutation";

export * from "./registerMutation";

export * from "./verifyEmailMutation";

export * from "./submitConsent";

export * from "./submitForgotPassword";

export * from "./submitPasswordChange";
