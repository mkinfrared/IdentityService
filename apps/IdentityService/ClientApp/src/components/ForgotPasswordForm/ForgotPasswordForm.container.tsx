import { memo } from "react";

import { useReturnUrl } from "hooks";

import ForgotPasswordFormFC from "./ForgotPasswordForm";
import { ForgotPasswordFormContainerProps } from "./ForgotPasswordForm.type";

const ForgotPasswordFormContainer = (
  props: ForgotPasswordFormContainerProps,
) => {
  const returnUrl = useReturnUrl();

  return <ForgotPasswordFormFC {...props} returnUrl={returnUrl} />;
};

export default memo(ForgotPasswordFormContainer);
