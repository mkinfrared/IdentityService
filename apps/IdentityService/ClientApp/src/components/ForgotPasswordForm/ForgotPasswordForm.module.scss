@import "../../styles/common";

.ForgotPasswordForm {
  @extend %loginRegisterForm;

  .successMessage {
    text-align: center;
  }
}

.Story {
}
