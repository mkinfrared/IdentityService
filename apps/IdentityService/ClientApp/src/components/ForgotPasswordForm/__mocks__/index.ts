export { default } from "./ForgotPasswordForm.mock";
