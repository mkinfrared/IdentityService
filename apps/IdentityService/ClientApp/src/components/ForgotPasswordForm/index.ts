export { default } from "./ForgotPasswordForm.container";

export * from "./ForgotPasswordForm.type";
