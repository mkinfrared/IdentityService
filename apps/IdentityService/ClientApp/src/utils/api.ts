import axios from "axios";

import { BASE_URL } from "utils/secrets";

const { isAxiosError } = axios;

const api = axios.create({
  baseURL: BASE_URL,
  withCredentials: true,
});

export { isAxiosError };

export default api;
