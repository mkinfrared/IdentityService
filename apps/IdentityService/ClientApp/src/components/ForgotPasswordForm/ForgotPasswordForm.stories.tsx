import { Meta, Story } from "@storybook/react";

import { ForgotPasswordForm } from "./ForgotPasswordForm";
import css from "./ForgotPasswordForm.module.scss";
import { ForgotPasswordFormProps } from "./ForgotPasswordForm.type";

export default {
  title: "COMPONENTS/ForgotPasswordForm",
  component: ForgotPasswordForm,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<ForgotPasswordFormProps> = (args) => (
  <div className={css.Story}>
    <ForgotPasswordForm {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {
  returnUrl: "/foo/bar",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
