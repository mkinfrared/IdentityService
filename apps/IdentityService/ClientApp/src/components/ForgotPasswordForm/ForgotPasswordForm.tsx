import { yupResolver } from "@hookform/resolvers/yup";
import { classNames } from "@identity-service/core";
import { Button, Heading } from "@identity-service/ui";
import { memo } from "react";
import { useForm } from "react-hook-form";

import { submitForgotPassword } from "api/mutations";
import { ForgotPasswordDto } from "api/types";
import InputField from "components/InputField";
import { ReactComponent as Email } from "icons/email-icon.svg";
import { isAxiosError } from "utils/api";
import { forgotPasswordSchema } from "utils/validationSchemas";

import css from "./ForgotPasswordForm.module.scss";
import { ForgotPasswordFormProps } from "./ForgotPasswordForm.type";

const ForgotPasswordForm = ({
  className,
  returnUrl,
}: ForgotPasswordFormProps) => {
  const { register, formState, handleSubmit, setError } =
    useForm<ForgotPasswordDto>({
      defaultValues: {
        email: "",
        returnUrl,
      },
      resolver: yupResolver(forgotPasswordSchema),
      mode: "onTouched",
    });

  const { errors, isSubmitting, isSubmitSuccessful } = formState;

  const onSubmit = async (values: ForgotPasswordDto) => {
    try {
      await submitForgotPassword(values);
    } catch (e) {
      if (isAxiosError(e)) {
        setError("email", { message: e.response?.statusText });
      }

      console.error(e);
    }
  };

  return (
    <div
      className={classNames(css.ForgotPasswordForm, className)}
      data-testid="ForgotPasswordForm"
    >
      <form className={css.form} onSubmit={handleSubmit(onSubmit)}>
        <Email className={css.icon} />
        <Heading className={css.heading}>Forgot Password</Heading>
        {isSubmitSuccessful ? (
          <Heading className={css.successMessage} variant="h4">
            Check your email
          </Heading>
        ) : (
          <>
            <InputField
              {...register("email")}
              autoComplete="off"
              error={errors.email?.message}
              placeholder="Email"
            />
            <div className={css.buttonWrapper}>
              <Button
                type="submit"
                disabled={isSubmitting}
                loading={isSubmitting}
              >
                Submit
              </Button>
            </div>
          </>
        )}
      </form>
    </div>
  );
};

export { ForgotPasswordForm };

export default memo(ForgotPasswordForm);
