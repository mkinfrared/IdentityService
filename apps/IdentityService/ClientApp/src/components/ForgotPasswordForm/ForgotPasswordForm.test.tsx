import { findByText, fireEvent, render, waitFor } from "@testing-library/react";

import * as mutations from "api/mutations";

import { ForgotPasswordForm } from "./ForgotPasswordForm";

jest.mock("api/mutations");

describe("<ForgotPasswordForm />", () => {
  const mutationsMock = mutations as jest.Mocked<typeof mutations>;
  const returnUrl = "/foo/bar";
  const Component = <ForgotPasswordForm returnUrl={returnUrl} />;

  it("should be defined", () => {
    expect(ForgotPasswordForm).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("ForgotPasswordForm");

    expect(element).toBeDefined();
  });

  it("should call submitForgotPassword on submit", async () => {
    mutationsMock.submitForgotPassword.mockResolvedValueOnce({});

    const data = {
      email: "timmy@coons.com",
      returnUrl,
    };

    const { getByTestId, container } = render(Component);
    const emailField = container.querySelector('input[name="email"]');
    const submitButton = getByTestId("Button");

    fireEvent.change(emailField!, { target: { value: data.email } });

    fireEvent.submit(submitButton);

    await waitFor(async () => {
      expect(mutationsMock.submitForgotPassword).toHaveBeenCalled();

      expect(mutationsMock.submitForgotPassword).toHaveBeenCalledTimes(1);

      expect(mutationsMock.submitForgotPassword).toHaveBeenCalledWith(data);

      const header = await findByText(container, /Check your email/i);

      expect(header).toBeVisible();
    });
  });

  it("should not call submitForgotPassword on submit when email is invalid", async () => {
    const data = {
      email: "timmy",
      returnUrl,
    };

    const { getByTestId, container } = render(Component);
    const emailField = container.querySelector('input[name="email"]');
    const submitButton = getByTestId("Button");

    fireEvent.change(emailField!, { target: { value: data.email } });

    fireEvent.submit(submitButton);

    await waitFor(() => {
      expect(mutationsMock.submitForgotPassword).not.toHaveBeenCalled();
    });
  });

  it("should show an error if submit failed", async () => {
    const message = "maklar";

    const responseMock = {
      isAxiosError: true,
      response: { statusText: message },
    };

    mutationsMock.submitForgotPassword.mockRejectedValueOnce(responseMock);

    const data = {
      email: "timmy@coons.com",
      returnUrl,
    };

    const { getByTestId, container } = render(Component);
    const emailField = container.querySelector('input[name="email"]');
    const submitButton = getByTestId("Button");

    fireEvent.change(emailField!, { target: { value: data.email } });

    fireEvent.submit(submitButton);

    await waitFor(async () => {
      expect(mutationsMock.submitForgotPassword).toHaveBeenCalled();

      const error = await findByText(container, message);

      expect(error).toBeVisible();
    });
  });
});
