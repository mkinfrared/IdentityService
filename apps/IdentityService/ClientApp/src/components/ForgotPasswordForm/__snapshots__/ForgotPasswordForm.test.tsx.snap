// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPasswordForm /> should match the snapshot 1`] = `
<div>
  <div
    class="ForgotPasswordForm"
    data-testid="ForgotPasswordForm"
  >
    <form
      class="form"
    >
      <svg
        class="icon"
      >
        email-icon.svg
      </svg>
      <h3
        class="Heading-module_Heading__Wg8yB heading"
        data-testid="Heading"
      >
        Forgot Password
      </h3>
      <label
        class="TextField-module_TextField__QtlJl"
        data-testid="TextField"
      >
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            autocomplete="off"
            name="email"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <div
        class="buttonWrapper"
      >
        <button
          class="Button-module_Button__0OJEu Button-module_opaque__O0DEF"
          data-testid="Button"
          type="submit"
        >
          <svg
            class=""
            fill="#fff"
            viewBox="0 0 135 140"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="120"
              rx="6"
              width="15"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="30"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="140"
              rx="6"
              width="15"
              x="60"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="90"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="120"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
          </svg>
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
